### ISSUE TITLE
data model

### ISSUE THREAD

**Issue by jtautry** [abruck1/APT2017-pink#2]
Morning gents!  It's a good day, I feel like we are going to get moving!  I have a writing session with the band till this afternoon and then I'll be heads down on this stuff till we meet.  I feel like a have a good handle on the data model but we need to make a couple design decisions.  One of which I'll pose to you now so we can talk about it this afternoon.

There are two ways to implement the tags functionality.  a) stream.tags - a repeated property of the stream class.  This is optimized for read/write of the entire list in one query.  The repeated part is how you store lists in ndb.  And I think it will be easier to use the Search API for full text search on this field across all stream entities.

The second would be a separate tag entity that would have the stream it belongs to as its parent.  This is closer to a normalized relational design, but in my opinion it has too much overhead.

So we can make this decision when we get together and go over my design of the data model.  If we all have time, I'd like to use the meeting to merge my changes back into Ari's branch, and then merge Ari's branch back into master so we have a solid base to finish up the work.  I would also like to go over how the dev environments are setup and make sure that each of us can run the app and crud data locally.

**Comment by abruck1** [comment]
Looks like we nailed this down. One issue left that I see:
Since we are not using ancestor keys, it looks like ndb is "eventually" consistent and not strongly consistent. This is a problem when deleting and unsubscribing from streams...any thoughts as to how to change our queries to handle this?

https://cloud.google.com/appengine/docs/standard/python/ndb/queries

Look at the Ancestor Queries section

**Comment by jtautry** [comment]
Is there a problem with eventually consistent?  That just means that the changes might not be reflected immediately across the geographically dispersed datastores.  We don't really care if a stream ends up not being deleted immediately to someone in china right?

**Comment by abruck1** [comment]
Yes, the problem is when you delete a stream or unsubscribe the page needs another refresh for the table to get updated

**Comment by abruck1** [comment]
we could put in a wait() but that seems silly

**Comment by jtautry** [comment]
that's not a problem with the consistency model though.  that's a databinding issue.

**Comment by jtautry** [comment]
we have to force a page refresh.

**Comment by jtautry** [comment]
This mock says to just reload the page:

https://apt.mybalsamiq.com/mockups/1083489.png?key=c6286db5bf27f95012252833d5214a336f17922c
